File: app/src/main/java/com/example/firefighters/viewmodels/UserViewModel.java
package com.example.firefighters.viewmodels;

import android.app.Activity;

import com.example.firefighters.models.UserModel;
import com.example.firefighters.repositories.UserRepository;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

public class UserViewModel extends ViewModel {

    private UserRepository repository;

    public void init() {
        if (repository != null)
            return;
        repository = UserRepository.getInstance();
    }

    public LiveData<QuerySnapshot> loadAllFireFightersWorkingUnitQuery(String unit){
        return repository.loadAllFireFightersWorkingUnitQuery(unit);
    }
    public LiveData<QuerySnapshot> loadFireFighters(DocumentSnapshot lastDocument, int limitCount){
        return repository.loadFireFightersQuery(lastDocument, limitCount);
    }
    public LiveData<QuerySnapshot> loadFireFightersSnapshot(){
        return repository.loadFireFightersQuerySnapshot();
    }
    public LiveData<Integer> saveFireFighter(UserModel firefighterModel) {
        return repository.saveFireFighter(firefighterModel);
    }
    public LiveData<Integer> updateFireFighter(UserModel firefighterModel) {
        return repository.updateFireFighter(firefighterModel);
    }
    public LiveData<Integer> deleteFireFighter(UserModel firefighterModel) {
        return repository.deleteFireFighter(firefighterModel);
    }

    //Sign in with mail and password
    public LiveData<Integer> signInUser(String userMail, String userPassword) {
        return repository.signInUser(userMail, userPassword);
    }
    //Create new user
    public LiveData<Integer> createNewUser(String userName, String userMail, String userPassword, boolean isFirefighter) {
        return repository.createNewUser(userName, userMail, userPassword, isFirefighter);
    }
    //Load to my page
    public LiveData<UserModel> loadUserModel(String userMail) {
        return repository.loadUserModel(userMail);
    }
    //Reset password
    public LiveData<Integer> resetPasswordWithMail(Activity activity, String userMail) {
        return repository.resetPasswordWithMail(activity, userMail);
    }
    //Disconnect user
    public LiveData<Integer> logOut() {
        return repository.logOut();
    }

}

File: app/src/main/java/com/example/firefighters/viewmodels/models/UserViewModel.kt
package com.example.firefighters.viewmodels.models

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.firefighters.models.UserModel
import com.example.firefighters.repositories.UserRepository
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.QuerySnapshot

class UserViewModel(private val mRepository: UserRepository) : ViewModel() {

    fun loadAllFireFightersWorkingUnitQuery(unit: String?): LiveData<QuerySnapshot> {
        return mRepository.loadAllFireFightersWorkingUnitQuery(unit)
    }

    fun loadFireFighters(
        lastDocument: DocumentSnapshot?,
        limitCount: Int
    ): LiveData<QuerySnapshot> {
        return mRepository.loadFireFightersQuery(lastDocument, limitCount)
    }

    fun loadFireFightersSnapshot(): LiveData<QuerySnapshot?> {
        return mRepository.loadFireFightersQuerySnapshot()
    }

    fun saveFireFighter(firefighterModel: UserModel?): LiveData<Int> {
        return mRepository.saveFireFighter(firefighterModel)
    }

    fun updateFireFighter(firefighterModel: UserModel?): LiveData<Int> {
        return mRepository.updateFireFighter(firefighterModel!!)
    }

    fun deleteFireFighter(firefighterModel: UserModel?): LiveData<Int> {
        return mRepository.deleteFireFighter(firefighterModel!!)
    }

    //Sign in with mail and password
    fun signInUser(userMail: String?, userPassword: String?): LiveData<Int> {
        return mRepository.signInUser(userMail, userPassword)
    }

    //Create new user
    fun createNewUser(
        userName: String?,
        userMail: String?,
        userPassword: String?,
        isFirefighter: Boolean
    ): LiveData<Int> {
        return mRepository.createNewUser(userName, userMail, userPassword, isFirefighter)
    }

    //Load to my page
    fun loadUserModel(userMail: String?): LiveData<UserModel?> {
        return mRepository.loadUserModel(userMail)
    }

    //Reset password
    fun resetPasswordWithMail(userMail: String?): LiveData<Int> {
        return mRepository.resetPasswordWithMail(userMail)
    }

    //Disconnect user
    fun logOut(): LiveData<Int> {
        return mRepository.logOut()
    }

    class Factory(
        private val repository: UserRepository
    ) : ViewModelProvider.NewInstanceFactory() {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return UserViewModel(repository) as T
        }
    }
}
